include/flux/op/cartesian_product.hpp

```diff
@@ -88,21 +89,36 @@ struct cartesian_product_traits_base {
 
     template <std::size_t I, typename Self>
     static constexpr auto ra_inc_impl(Self& self, cursor_type<Self>& cur, distance_t offset)
-    -> cursor_type<Self>&
+        -> cursor_type<Self>&
     {
+        if (offset == 0)
+            return cur;
+
         auto& base = std::get<I>(self.bases_);
+        auto& this_index = std::get<I>(cur);
+        auto this_size = flux::size(base);
 
-        auto this_sz = flux::size(base);
-        auto this_offset = offset  % this_sz;
-        auto next_offset = offset/this_sz;
+        this_index += offset;
```

Comment:
Use `checked_add` here to avoid potential UB on signed overflow

```suggestion
        this_index = num::checked_add(this_index, offset);
```

(Yeah, it's ugly, I know :( )

Reply:
Fixed in ba910d7c432ec5b70ba58ed428b1cbbb48360c25.